# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [205]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier


## 1. Loading the dataset

In [201]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
slime = pd.read_csv("science_data_large.csv")
print(slime.head(15))
print(slime.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [202]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = slime.drop(columns=["Size nm^3"])
y = slime["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9,test_size=0.1, random_state=42)
# For grading consistency use random_state=42 
print(X_train)
print(X_test)
print(y_train)
print(y_test)

     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]
     Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              580       817
826              24        89

[100 rows x 2 columns]
716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size n

## 3. Perform a Linear Regression

In [203]:
# Use sklearn to train a model on the training set
m = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
datap = X_train.head(1)
p_out = m.predict(datap)
print(p_out)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
m_score = m.score(X_test, y_test)
print(m_score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = m.coef_
intercept = m.intercept_
print(coefficients)
print(intercept)

[188207.26786334]
0.8552472077276095
[ 866.14641337 1032.69506649]
-409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)
\begin{equation}
h(x) = 866.15x_1 + 1032.70x_2 - 409390.0
\end{equation}

Report on score and explain meaning: The score is 0.85525, this mean that the model is fairly close to the acutal average of the data and this means its result are fairly cloe to the actual values.

## 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
model_cross = LinearRegression()
cv = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(model_cross, X, y, cv=cv)
# Report on their finding and their significance
print(scores)


[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here: [0.86152 0.82742 0.87195 0.88166 0.85609], after randomly splitting the orginal data 5 times into differnt sets, the number returned is R^2, or our score from the earlier question. Meaning this is how close the given set is to the mean and then meaning how accurate it is predicating a value. So the fact that our data is consinetly > 8 means the model is fairly accurate. 

## 5. Using Polynomial Regression

In [208]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
cv = KFold(n_splits=5, shuffle=True, random_state=42)
#X_test_poly = poly.transform(X_test)
score = []

for train_i, test_i in cv.split(X_train):
    X_train_fold, X_test_fold = X_train.iloc[train_i], X_train.iloc[test_i]
    y_train_fold, y_test_fold = y_train.iloc[train_i], y_train.iloc[test_i]
    poly = PolynomialFeatures(degree=2)
    X_train_poly = poly.fit_transform(X_train_fold)
    X_test_poly = poly.transform(X_test_fold)
    model = LinearRegression().fit(X_train_poly, y_train_fold)
    score.append(model.score(X_test_poly,y_test_fold))
    
    
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X_train)
model = LinearRegression().fit(X_poly,y_train)

# Report on the metrics and output the resultant equation as you did in Part 3.
coefficients = model.coef_
intercept = model.intercept_
print("Coefficients:", coefficients)
print("Intercept:", intercept)
print("Cross-validated R² scores:", score)
print("Mean R² score:", np.mean(score))

Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05
Cross-validated R² scores: [1.0, 1.0, 1.0, 1.0, 1.0]
Mean R² score: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

\begin{equation}
h(x) = 2.0477 \times 10^{-5} + 12.00 x_1 - 1.27200 \times 10^{-7} x_2 + 1.2649 \times 10^{-11} x_3 + 2.00 x_4 + 0.028571 x_5

\end{equation}

Report on the score and interpret: [1.0, 1.0, 1.0, 1.0, 1.0] are our r^2 scores. This could be a gross overfitting from our model since it is telling us we are perfect nad have a model that will always give the corect data. So their most be some sort of overfitting going on with the data for it to be so accurate.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [129]:
# Load the dataset. Then train and evaluate the classification models.
clean_ckd = pd.read_csv("ckd_feature_subset.csv")
X = clean_ckd.drop(columns=["Target_ckd"])
y = clean_ckd["Target_ckd"]
kf = KFold(n_splits=5, shuffle=True, random_state=42)
# X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9,test_size=0.1, random_state=42)



# model = LogisticRegression().fit(X_train,y_train)
# lin_reg_score = cross_val_score(model, X_train, y_train, cv=kf)
model = LogisticRegression()
lin_reg_score = cross_val_score(model, X,y, cv=kf)
lin_reg_mean = lin_reg_score.mean()
lin_reg_sd = lin_reg_score.std()
print("LogisticRegression")
print(lin_reg_mean)
print(lin_reg_sd)


# model = SVC().fit(X_train,y_train)
# svc_score = cross_val_score(model, X_train, y_train, cv=kf)
model = SVC()
svc_score = cross_val_score(model, X, y, cv=kf)
svc_mean = svc_score.mean()
svc_sd = svc_score.std()
print("SVC")
print(svc_mean)
print(svc_sd)


# model = KNeighborsClassifier().fit(X_train,y_train)
# kneigh_score = cross_val_score(model, X_train, y_train, cv=kf)
model = KNeighborsClassifier()
kneigh_score = cross_val_score(model, X, y, cv=kf)
kneigh_score_mean = kneigh_score.mean()
kneigh_score_sd = kneigh_score.std()
print("KNeighborsClassifier")
print(kneigh_score_mean)
print(kneigh_score_sd)


# model = MLPClassifier(max_iter=1000).fit(X_train,y_train)
# mlp_score = cross_val_score(model, X_train, y_train, cv=kf)
model = MLPClassifier(max_iter=1000)
mlp_score = cross_val_score(model, X, y, cv=kf)
mlp_mean = mlp_score.mean()
mlp_sd = mlp_score.std()
print("MLPClassifier")
print(mlp_mean)
print(mlp_sd)

LogisticRegression
0.8565591397849461
0.0662686697755876
SVC
0.9281720430107526
0.04760126519964333
KNeighborsClassifier
0.9279569892473118
0.05243961560473918
MLPClassifier
0.9350537634408603
0.040465552982107675


## Results and Conclusion for Classification Experiments

| Model Name | Mean | Std |
|------------------------|----------------------|------------------------|
| Logistic Regression | 0.85656 | 0.066269 |
| SVC | 0.92827 | 0.047601 |
| KNeighbors Classifier | 0.92797 | 0.052440 |
| MLP Classifier | 0.93505 | 0.040466 |

Interperting Meaning:
The best performing model was the MLP Classfier with the highest standard R^2 mean and the lowest standard devation among the scores. This means it as a high degree of accuracy and stability across any split of the data. Both SVC and KNeighbors both had a nearly identical mean and vaired with their STD. So both perforemd really well across multiple diffenrt folds. Logistic Regeression performed with the worst the lowest mean and high STD, so in comparsion to the other did was not as accurate and had a wider range of instaibtily. MLP may perform the best since it is a neural netowrk allowing it to capture more complex and non linear boundires within the dataset. 


In [130]:
# Experiments with Neural Network.
# model = MLPClassifier(max_iter=1000,hidden_layer_sizes=(200,)).fit(X_train,y_train)
# mlp_score1 = cross_val_score(model, X_test, y_test, cv=kf)
model = MLPClassifier(max_iter=2000,hidden_layer_sizes=(200,))
mlp_score1 = cross_val_score(model, X, y, cv=kf)
mlp_mean = mlp_score1.mean()
mlp_sd = mlp_score1.std()
print("MLPClassifier1")
print(mlp_mean)
print(mlp_sd)

# model = MLPClassifier(max_iter=1000,activation='logistic').fit(X_train,y_train)
# mlp_score2 = cross_val_score(model, X_test, y_test, cv=kf)
model = MLPClassifier(max_iter=2000,activation='logistic')
mlp_score2 = cross_val_score(model, X, y, cv=kf)
mlp_mean = mlp_score2.mean()
mlp_sd = mlp_score2.std()
print("MLPClassifier2")
print(mlp_mean)
print(mlp_sd)


# model = MLPClassifier(max_iter=1000,solver='lbfgs').fit(X_train,y_train)
# mlp_score3 = cross_val_score(model, X_test, y_test, cv=kf)
model = MLPClassifier(max_iter=2000,solver='lbfgs')
mlp_score3 = cross_val_score(model, X, y, cv=kf)
mlp_mean = mlp_score3.mean()
mlp_sd = mlp_score3.std()
print("MLPClassifier3")
print(mlp_mean)
print(mlp_sd)

MLPClassifier1
0.9350537634408603
0.040465552982107675
MLPClassifier2
0.9217204301075268
0.04835315416425381
MLPClassifier3
0.9610752688172044
0.031565321517262414


## Results and Conclusion for Neural Network Experiments

| Model Name | Mean | Std |
|------------------|----------|----------|
| MLP Classifier 1 | 0.93505 | 0.04047 |
| MLP Classifier 2 | 0.91527 | 0.04824 |
| MLP Classifier 3 | 0.98043 | 0.02592 |

In the first example we changed the number of hidden layers which did not change anything in comparssion to the orgianl run of the MLP in the one before. Meaning that change the number of hidden layers had little to no effect on improving the model. In the 2nd example I change the activation function to logistic, which is differnt then the deafult relu. This made the mdoel perform worse with a lower mean and a highert Std. The one that imporved and had the largest affect on the mean was changing the solver to lbfgs in the 3rd example. The mean had the largest positive imporvement and a larger decrease in std. Meaning that we increased the accuracy and stbaitlity of the model by change the solver parameter. 